{"cells":[{"cell_type":"markdown","metadata":{"id":"VZpWDT_qK5mE"},"source":["# Module 4: Variant Lineage Identification\n","\n","Welcome to the module! There are some very important instructions for you to follow:\n","\n","1.) Click on File on the top left corner and select save a copy in drive\n","\n","**Your changes will not be saved if you do not do this step**\n","\n","2.) Click on the name of the workbook in the top left corner and replace \"Copy of\" with your full name"]},{"cell_type":"markdown","metadata":{"id":"G7mk1VecOHLQ"},"source":["## Setting up\n","\n","## Installing Conda\n","Conda is a versatile software management tool. Conda is an open source system of managing tools and libraries. More info on the library used to install conda on Google Colab is at this [website](https://inside-machinelearning.com/en/how-to-install-use-conda-on-google-colab/)\n","\n","Note - your runtime will refresh and reconnect after running this. It will say runtime crashed, this seems normal, wait for the session to reconnect after this.\n","\n","You can check out this repo for how this tool works:\n","https://github.com/conda-incubator/condacolab \n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jOTBUZ3gOFm3"},"outputs":[],"source":["!pip install -q condacolab\n","import condacolab\n","condacolab.install()"]},{"cell_type":"markdown","metadata":{"id":"yESC04s9KoEd"},"source":["## Testing your conda installation\n","After successfully install condacolab, you need to be sure that you can call conda from your shell and install software and also execute them. \n","The `--help` command is an easy way to test that it is properly installed as it will display the help options used in running the tool."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"flkQBFR1K892"},"outputs":[],"source":["#You can check conda installation by launching the help command\n","!conda --help\n"]},{"cell_type":"markdown","metadata":{"id":"XxSaFl6slUXM"},"source":["## Check pre-installed conda packages\n","After conda is installed, it comes with some packages and software that are installed alongside the software. To generate a list of these packages, simply run the code below:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"h00eUuhdlwy_"},"outputs":[],"source":["!conda list"]},{"cell_type":"markdown","metadata":{"id":"QYH__ysqA8Tw"},"source":["## Download and navigate to directory containing genomes\n","There are some test SARS-CoV-2 genome assemblies obtained from GISAID, saved in the file `assemblies_gisaid.zip`. Download and unzip the file, which will create a folder called `assemblies_gisaid`. Then use the `cd` command to switch to that directory:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"RKLApCAaBbx3"},"outputs":[],"source":["!wget https://zenodo.org/records/10888461/files/module_3.tar"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["!tar -xvf module_3_P1.tar"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"CoMgBYUecoAF"},"outputs":[],"source":["%cd module_3_P1/\n","!pwd"]},{"cell_type":"markdown","metadata":{},"source":["The data analyzed in this module are derived from the article \"[Overview of the SARS-CoV-2 genotypes circulating in Latin America during 2021]( https://doi.org/10.3389/fpubh.2023.1095202 )\" published by Molina-Mora et al. in 2023, as part of a CABANA project.\n","Below, you will find the Accession ID from the GISAID platform and the country from which the sample originated:"]},{"cell_type":"markdown","metadata":{},"source":["#### Genomes \n","| Country     | Accession ID   |\n","|-------------|----------------|\n","| Argentina   | EPI_ISL_14434222 |\n","| Argentina   | EPI_ISL_14434402 |\n","| Argentina   | EPI_ISL_14434358 |\n","| Bolivia     | EPI_ISL_8800564  |\n","| Bolivia     | EPI_ISL_8800607  |\n","| Bolivia     | EPI_ISL_8800591  |\n","| Costa Rica  | EPI_ISL_7711628  |\n","| Costa Rica  | EPI_ISL_7711763  |\n","| Costa Rica  | EPI_ISL_7711812  |\n","| Colombia    | EPI_ISL_10072006 |\n","| Colombia    | EPI_ISL_10072040 |\n","| Colombia    | EPI_ISL_10080397 |\n","| Mexico      | EPI_ISL_7812926  |\n","| Mexico      | EPI_ISL_7812869  |\n","| Mexico      | EPI_ISL_7813015  |\n","| Peru        | EPI_ISL_7961355  |\n","| Peru        | EPI_ISL_7961418  |\n","| Peru        | EPI_ISL_7961482  |\n","| Brazil      | EPI_ISL_3369834  |\n","| Brazil      | EPI_ISL_3369992  |\n","| Brazil      | EPI_ISL_3373439  |"]},{"cell_type":"markdown","metadata":{"id":"qk2HjgKTDCpA"},"source":["# Pangolin\n","\n","PANGOLIN is an acronym for Phylogenetic Assignment of Named Global Outbreak Lineages. It is a software that assigns your SARS-CoV-2 genome to the most closest related SARS-CoV-2 lineage on the global context based on the mutations in the query sequence. Pangolin can be accessed on its web application which allows you to upload your FASTA files here: https://pangolin.cog-uk.io/. "]},{"cell_type":"markdown","metadata":{"id":"O8a71SiXma5u"},"source":["### Installing Pangolin for variant detection\n","Pangolin can be installed within conda by running the code below:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"FKkeS08Izq1e"},"outputs":[],"source":["!conda install -c bioconda -c conda-forge -c defaults pangolin"]},{"cell_type":"markdown","metadata":{"id":"kz78Z-Y37Y09"},"source":["To check that Pangolin installed correctly, run the following command:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"C04BdH67701K"},"outputs":[],"source":["!pangolin -h"]},{"cell_type":"markdown","metadata":{"id":"uJcGpAWjMeXS"},"source":["### Launching Pangolin on the multi-FASTA file\n","Now you have successfully installed Pangolin and displayed the help options which shows you the commands and options you can adjust to execute your analysis with the software.\n","\n","We will now try to identify the variants of the genomes in the merged FASTA file using Pangolin with the default settings; and direct our output to the `results` folder. By default, the output file will be called `lineage_report.csv`."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"EM2LA7s1hn_e"},"outputs":[],"source":["!pangolin --outdir results gisaid_hcov-19_2024_03_27_02.fasta"]},{"cell_type":"markdown","metadata":{"id":"BYSGi1aaQH9e"},"source":["**NOTE**: Download the `lineage_report.csv` file to your computer and open it with a Spreadsheet software like MS Excel."]},{"cell_type":"markdown","metadata":{"id":"EUFQHiEmQm3D"},"source":["# Nextclade\n","\n","Nextclade is another fantastic software that can identify variants in your genomes like Pangolin. However, it has additional features in its report, like the number of mutations across the genome and where they are located. Nextclade can be accessible on the web application: https://clades.nextstrain.org/."]},{"cell_type":"markdown","metadata":{"id":"wtweiWDmEdbI"},"source":["### Installing Nextclade\n","\n","Install Nextclade with conda using the following command, and then test the installing by displaying the help page."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"L6Md_ww7Ro1Z"},"outputs":[],"source":["!conda install -c bioconda nextclade"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"f0o_RYdASDTt"},"outputs":[],"source":["# To check if Nextclade insalled properly\n","!nextclade --help"]},{"cell_type":"markdown","metadata":{"id":"AqF4j9ItEzjw"},"source":["### Running NextClade\n","\n","Before running Nextclade on the assembled genomes, first download the latest available Nextclade dataset for SARS-CoV-2."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0vu235maYaZG"},"outputs":[],"source":["# Download the SARS-CoV-2 dataset folder for NextClade\n","!nextclade dataset get --name sars-cov-2 --output-dir sars-cov-2-dataset"]},{"cell_type":"markdown","metadata":{"id":"oZ1KkR8XQo_g"},"source":["Then identify the clades from the genomes in the multi-FASTA file."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"kXAEnLGdSlkF"},"outputs":[],"source":["# Run NextClade on your concatenated FASTA files to generate TSV output file\n","!nextclade run --output-tsv nextclade_report.tsv --input-dataset sars-cov-2-dataset gisaid_hcov-19_2024_03_27_02.fasta"]},{"cell_type":"markdown","metadata":{"id":"PgjfKUueaSBH"},"source":["**NOTE**: Download the \"nextclade_report.tsv\" file and open it with a spreadsheet software on your computer like MS Excel."]},{"cell_type":"markdown","metadata":{"id":"IxwWWeddagN0"},"source":["## Assignment\n","\n","1. From your NextClade results, what is the most persistent clade in the samples?\n","\n","2. Also from your NextClade results, how many amino acid substitutions does the genome EPI_ISL_7711812 have?\n","\n","3. What is the genome with the highest number of \"missing\" entries?"]}],"metadata":{"colab":{"provenance":[],"toc_visible":true},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
